### PR TITLE
[new command] Get-DbaTraceFlag

### PR DESCRIPTION
**Pull request by sqlpadawan**
## Type of Change

 - [ ] Bug fix (non-breaking change, fixes #<enter issue number>)
 - [x] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Pester test is included

### Purpose
 
Poll sql instance global enabled trace flags from the startup parameters or from DBCC TRACEON(<tf>,-1)

### Approach

SMO trace flag enumeration

### Commands to test

examples included from the scripts:
.EXAMPLE
Get-DbaTraceFlag -SqlInstance localhost
Returns all SQL Trace Flag information on the local default SQL Server instance

.EXAMPLE
Get-DbaTraceFlag -SqlInstance localhost, sql2016
Returns all SQl Trace Flag for the local and sql2016 SQL Server instances

.EXAMPLE
Get-DbaTraceFlag -SqlInstance localhost -TraceFlag 4199,3205
Returns all SQl Trace Flag 4199 and 3205 status for local SQL Server instance

### Screenshots

capture1

### CODE REVIEW

File: functions/Get-DbaTraceFlag.ps1 [hunk 1]
```diff
@@ -0,0 +1,96 @@
+﻿FUNCTION Get-DbaTraceFlag {
+	<#
+		.SYNOPSIS
+			Gets SQL Trace Flags information for each instance(s) of SQL Server.
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: Would just say `Get global Trace Flag(s) information for each instance(s) of SQL Server.`

File: functions/Get-DbaTraceFlag.ps1 [hunk 2]
```diff
@@ -0,0 +1,96 @@
+﻿FUNCTION Get-DbaTraceFlag {
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: Please change to `function`

File: functions/Get-DbaTraceFlag.ps1 [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	<#
+		.SYNOPSIS
+			Gets SQL Trace Flags information for each instance(s) of SQL Server.
+
+		.DESCRIPTION
+			The Get-DbaTraceFlag returns connected SMO object for SQL Trace Flags information for each instance(s) of SQL Server.
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: No need to repeat the command name, can just say command or maybe reword to:

> Returns Trace Flags that are enabled globally on each instance(s) of SQL Server as an object.

Since it does not return an SMO just saying "object" is fine. As well, we need to call out that this only gets global trace flags...I know should be obvious but best to state it.

File: functions/Get-DbaTraceFlag.ps1 [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		foreach ($instance in $SqlInstance) {
+			Write-Verbose "Attempting to connect to $instance"
+			try {
+				$server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SqlCredential
+			}
+			catch {
+				Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+			}
+
+            $tflags = New-Object System.Data.DataTable;
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: This line can be removed.

File: functions/Get-DbaTraceFlag.ps1 [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		[DbaInstanceParameter[]]$SqlInstance,
+		[PSCredential]
+		$SqlCredential,
+		[object[]]$TraceFlag,
+		[switch]$Silent
+	)
+
+	process {
+		foreach ($instance in $SqlInstance) {
+			Write-Verbose "Attempting to connect to $instance"
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: This can be removed.

Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: We don't generally provide much output for Get commands even as verbose.

File: functions/Get-DbaTraceFlag.ps1 [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                               'InstanceName' = $server.ServiceName;
+                               'SqlInstance' = $server.DomainInstanceName;
+                               'TraceFlag' = $tflag.TraceFlag;
+                               'Global' = $tflag.Global;
+                               'Session' = $tflag.Session;
+                               'Status' = $tflag.Status};
+                               
+                $tflagrow = New-Object psobject -Property $tflagdata
+                
+                Select-DefaultView -InputObject $tflagrow -Property 'ComputerName','InstanceName','SqlInstance','TraceFlag','Global','Session','Status'
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: We can clean this up a bit as utilizing power in version PS 3 for custom objects.

```powershell
			foreach ($tflag in $tflags) {
				[pscustomobject]@{
					'ComputerName' = $server.NetName;
					'InstanceName' = $server.ServiceName;
					'SqlInstance'  = $server.DomainInstanceName;
					'TraceFlag'    = $tflag.TraceFlag;
					'Global'       = $tflag.Global;
					'Session'      = $tflag.Session;
					'Status'       = $tflag.Status
				} | Select-DefaultView -ExcludeProperty 'Session'
			}
```
This is only because I've seen @potatoqualitee do this shortcut method in one of the commands I contributed 💯 

Being that Session is always going to be zero, not much reason to include it in the default layout to me.

File: functions/Get-DbaTraceFlag.ps1 [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+			try {
+				$server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SqlCredential
+			}
+			catch {
+				Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+			}
+
+            $tflags = New-Object System.Data.DataTable;
+
+			$tflags = $server.EnumActiveGlobalTraceFlags();
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: I'll leave it to @potatoqualitee for final vote but I think this command deserves some output if there are no global TFs found.

Something like

```powershell
			if ($tFlags.Rows.Count -eq 0) {
				Write-Message -Level Output -Message "No global trace flags enabled"
				return
			}
```

File: functions/Get-DbaTraceFlag.ps1 [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+			Use this switch to filter to a specific Trace Flag.
+
+		.PARAMETER Silent
+			Use this switch to disable any kind of verbose messages.
+
+		.NOTES
+			Tags: Trace, Flag
+			Original Author: Kevin Bullen (@sqlpadawan)
+
+			References:  https://msdn.microsoft.com/en-us/library/microsoft.sqlserver.management.smo.server.enumactivecurrentsessiontraceflags(v=sql.120).aspx
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: We don't generally provide links to the SMO object utilized, this would be more to note in your PR.

If you would update that URL to this one: `https://docs.microsoft.com/en-us/sql/t-sql/database-console-commands/dbcc-traceon-trace-flags-transact-sql`

File: functions/Get-DbaTraceFlag.ps1 [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			Copyright: (C) Chrissy LeMaire, clemaire@gmail.com
+			License: GNU GPL v3 https://opensource.org/licenses/GPL-3.0
+
+		.LINK
+			https://dbatools.io/Get-DbaTraceFlag
+
+		.EXAMPLE
+			Get-DbaTraceFlag -SqlInstance localhost
+
+			Returns all SQL Trace Flag information on the local default SQL Server instance
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: Can remove `SQL` and just state `Trace Flag`

File: functions/Get-DbaTraceFlag.ps1 [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+		.EXAMPLE
+			Get-DbaTraceFlag -SqlInstance localhost, sql2016
+
+			Returns all SQl Trace Flag for the local and sql2016 SQL Server instances
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: Can remove `SQL` and just state `Trace Flag`

File: functions/Get-DbaTraceFlag.ps1 [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+		.EXAMPLE
+			Get-DbaTraceFlag -SqlInstance localhost -TraceFlag 4199,3205
+
+			Returns all SQl Trace Flag 4199 and 3205 status for local SQL Server instance
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: I would word this as something like:

> Returns Trace Flag status for TF 4199 and 3205 for the local SQL Server instance if they are enabled.

File: functions/Get-DbaTraceFlag.ps1 [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+			catch {
+				Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+			}
+
+            $tflags = New-Object System.Data.DataTable;
+
+			$tflags = $server.EnumActiveGlobalTraceFlags();
+			
+			if ($TraceFlag) {
+				$tflags = $tflags | Where-Object TraceFlag -In $TraceFlag
```
Comment (wsmelton) [inline functions/Get-DbaTraceFlag.ps1]: 💯

**Comment by wsmelton** [comment]
@potatoqualitee I have a test created for this command under [this branch](https://github.com/sqlcollaborative/dbatools/tree/GetDbaTraceFlag-Tests)

**Comment by wsmelton** [comment]
@sqlpadawan I would clarify one thing in your PR statement:

> Enumerate trace flags from startup parameters

This does not necessarily enumerate startup parameters. It is only polling the SQL Server instance for globally enabled trace flags. Which those can be from startup parameters or from `DBCC TRACEON(<tf>,-1)` being executed.

File: functions/Get-DbaTraceFlag.ps1 (hunk 2 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 1 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 3 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 8 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted, this was for my notes, if needed it can be removed.

File: functions/Get-DbaTraceFlag.ps1 (hunk 9 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 10 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 11 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 5 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes submitted

File: functions/Get-DbaTraceFlag.ps1 (hunk 4 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: removed.

File: functions/Get-DbaTraceFlag.ps1 (hunk 7 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: i went and added the snippet (since you went to the trouble of writing it for me).

File: functions/Get-DbaTraceFlag.ps1 (hunk 12 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: not sure what the comment was here, sorry.

File: functions/Get-DbaTraceFlag.ps1 (hunk 6 above)
Comment (sqlpadawan) [inline functions/Get-DbaTraceFlag.ps1]: changes made.

**Comment by gbargsley** [comment]
All tests I ran worked.  I have a suggestion on the output when no global trace flags enabled.  In the top test I have two of those servers do not have trace flags on.  But from the output I guess you have to assume only the one that returns does.  Not sure if the message can show for each server listed that has no trace flags.

image


Other than this question I will approve.

**Comment by wsmelton** [comment]
I can add the server name to the output on the PR I have for the test on this command.

**Comment by potatoqualitee** [comment]
Really excited about this! Thanks to you all and welcome to the team @sqlpadawan! Will be adding your information to [dbatools.io/team](https://dbatools.io/team)

**Comment by sqlpadawan** [comment]
@gbargsley I did see your comment, and I see the problem. I'm testing a fix now, adding the server name to the output is easy enough.  But I also want to add a data to the pscustomobject to show the server was checked and there aren't any global trace flags set. Something like this:

            if ($tFlags.Rows.Count -eq 0) {
				Write-Message -Level Output -Message "$server has no global trace flags enabled"
                [pscustomobject]@{'ComputerName' = $server.NetName;
                		          'InstanceName' = $server.ServiceName;
                  		          'SqlInstance'  = $server.DomainInstanceName;
                  		          'TraceFlag'    = 0
                } | Select-DefaultView 
				return
			}

 I'm trying to replace manually created\maintained configuration spreadsheets with the various functions, and I'm throwing everything in a db right now. I want a record to show the server was checked and has no trace flags set.

**Comment by wsmelton** [comment]
We don't output an object if nothing is returned.

**Comment by wsmelton** [comment]
Rather if nothing is found on the given server then an output message is all we do in commands. 

Folks that want to populate the result of the command into a variable can no longer check for null if you output an object.

**Comment by sqlpadawan** [comment]
Ok, can you point me in the right direction for doing something like this:

Get-DbaTraceFlag -SqlInstance Win7X64Dev,Win7X64Dev\InstanceA | Out-DbaDataTable | Write-DbaDataTable -SqlServer . -Table tempdb.dbo.DbaTraceFlag -Truncate -AutoCreateTable -Confirm:$false -WarningAction:silentlycontinue

and still get a record for a server that doesn't have any trace flags set?  Or i suppose really at that point it could just be a left join to the trace flag table. But that's the difference between a record of nothing is there versus no records so nothing must be there.
